src/analyses/constant_propagator.cpp

```diff
@@ -231,42 +223,79 @@ bool constant_propagator_domaint::two_way_propagate_rec(
   std::cout << "two_way_propagate_rec: " << format(expr) << '\n';
 #endif
 
-  bool change=false;
+  // This used to attempt to propagate equalities both ways, but the original
+  // author noted it is "buggy at the moment," so I only propagate constants
+  // onto non-constants.
 
-  // this seems to be buggy at present
-#if 0
   if(expr.id()==ID_and)
   {
-    // need a fixed point here to get the most out of it
-    do
+    forall_operands(it, expr)
+      two_way_propagate_rec(*it, ns, cp);
+  }
+  else if(expr.id() == ID_not)
+  {
+    if(expr.op0().id() == ID_equal || expr.op0().id() == ID_notequal)
     {
-      change = false;
-
-      forall_operands(it, expr)
-        if(two_way_propagate_rec(*it, ns, cp))
-          change=true;
+      exprt subexpr = expr.op0();
+      subexpr.id(subexpr.id() == ID_equal ? ID_notequal : ID_equal);
+      two_way_propagate_rec(subexpr, ns, cp);
+    }
+    else if(expr.op0().id() == ID_symbol && expr.type() == bool_typet())
+    {
+      // Treat `IF !x` like `IF x == FALSE`:
+      two_way_propagate_rec(equal_exprt(expr.op0(), false_exprt()), ns, cp);
     }
-    while(change);
   }
-  else if(expr.id()==ID_equal)
+  else if(expr.id() == ID_symbol)
   {
-    const exprt &lhs=expr.op0();
-    const exprt &rhs=expr.op1();
-
-    // two-way propagation
-    valuest copy_values=values;
-    assign_rec(copy_values, lhs, rhs, ns);
-    if(!values.is_constant(rhs) || values.is_constant(lhs))
-       assign_rec(values, rhs, lhs, ns);
-    change=values.meet(copy_values);
+    if(expr.type() == bool_typet())
+    {
+      // Treat `IF x` like `IF x == TRUE`:
+      two_way_propagate_rec(equal_exprt(expr, true_exprt()), ns, cp);
+    }
   }
-#endif
+  else if(expr.id() == ID_equal || expr.id() == ID_notequal)
+  {
+    exprt lhs = expr.op0();
+    exprt rhs = expr.op1();
 
-#ifdef DEBUG
-  std::cout << "two_way_propagate_rec: " << change << '\n';
-#endif
+    if(lhs.is_constant() && !rhs.is_constant())
+      std::swap(lhs, rhs);
+
+    if(!(rhs.is_constant() && !lhs.is_constant()))
```

Comment:
I think it'd be clearer to expand the not here: `lhs.is_constant() || !rhs.is_constant()`.